### PR TITLE
1.2.1

### PR DESCRIPTION
**Pull request by Glitchfix**
What this PR does / why we need it:
PX-Terraform IBM - Scaling up storage nodes using TF, related to "max_storage_nodes_per_zone"

Which issue(s) this PR fixes (optional)
Closes #[PWX-27839](https://portworx.atlassian.net/browse/PWX-27839)

Special notes for your reviewer:
this PR contains code for autopilot installation code but is not part of this PR as more improvement on autopilot rules is required

### CODE REVIEW

**Review by vinayakshnd — CHANGES_REQUESTED** [review]
@Glitchfix pls remove autopilot related changes from this branch. Mergeing this branch will make feature available to users without testing. File a separate PR for autopilot

File: utils/portworx_configure_max_storage_node_per_zone.sh [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    tar -xzf /tmp/helm3/helm-v3.3.0-linux-amd64.tar.gz -C /tmp/helm3/
+    CMD="/tmp/helm3/linux-amd64/helm"
+    $CMD version
+fi
+
+# Get the portworx chart namespace
+CHART_NAMESPACE=$(helm ls -A | grep portworx | awk '{print $2}')
+# Get the Helm status
+if ! JSON=$(helm history portworx -n ${CHART_NAMESPACE} -o json | jq '. | last'); then
+    printf "[ERROR] Helm couldn't find Portworx Installation, will not proceed with the upgrade!! Please install portworx and then try to upgrade.\n"
```
Comment (vinayakshnd) [inline utils/portworx_configure_max_storage_node_per_zone.sh:44]: Change above ERROR msg.
```suggestion
    printf "[ERROR] Helm couldn't find Portworx Installation, will not proceed with configuring max storage nodes per zone!!\n"
```

File: utils/portworx_configure_max_storage_node_per_zone.sh [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        sleep $SLEEP_TIME
+    fi
+done
+
+# Check if portworx ds is there, if there,  get the ds details else, exit with error
+# Store the number of desired and ready pods
+# Show the current pods and ds status
+printf "[INFO] Validating Portworx Cluster Status...\n"
+if ! sc_state=$(kubectl get storagecluster ${PX_CLUSTER_NAME} -n ${NAMESPACE}); then
+    printf "[ERROR] Portworx Storage Cluster Not Found, will not proceed with the upgrade!! Please install Portworx Enterprise and then try to upgrade.\n"
```
Comment (vinayakshnd) [inline utils/portworx_configure_max_storage_node_per_zone.sh:72]: Same as above. Change error msg.

File: utils/portworx_configure_max_storage_node_per_zone.sh [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    exit 1
+else
+    STATUS=$(kubectl get storagecluster ${PX_CLUSTER_NAME} -n ${NAMESPACE} -o jsonpath='{.status.phase}')
+    if [ "$STATUS" != "Online" ]; then
+        printf "[ERROR] Portworx Storage Cluster is not Online. Cluster Status: ($STATUS), will not proceed with the upgrade!!\n"
```
Comment (vinayakshnd) [inline utils/portworx_configure_max_storage_node_per_zone.sh:77]: Same here

File: utils/portworx_configure_max_storage_node_per_zone.sh [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+fi
+echo "maxStorageNodesPerZone: ${NODE_COUNT}" >> $HELM_VALUES_FILE
+printf "[INFO] maxStorageNodesPerZone value updated in ${HELM_VALUES_FILE}!!\n"
+printf "[INFO] Upgrading ${HELM_VALUES_FILE}!!\n"
+$CMD upgrade portworx ibm-helm/portworx -f ${HELM_VALUES_FILE} -n ${CHART_NAMESPACE}
+
+if [[ $? -eq 0 ]]; then
+    printf "[INFO] Upgrade Triggered Succesfully, will monitor the storage cluster!!\n"
+else
+    printf "[ERROR] Failed to Upgrade!!\n"
```
Comment (vinayakshnd) [inline utils/portworx_configure_max_storage_node_per_zone.sh:134]: Upgrade refers to PX version upgrade. Replace `upgrade` keyword everywhere in the file with `configure maxStorageNodesPerZone`.

**Review by vinayakshnd — CHANGES_REQUESTED** [review]

**Comment by Glitchfix** [comment]
> @Glitchfix pls remove autopilot related changes from this branch. Mergeing this branch will make feature available to users without testing. File a separate PR for autopilot

the autopilot changes won't be triggered unless there is a flag set for it as the deffault value for it false

**Comment by vinayakshnd** [comment]
@Glitchfix if there is variable/flag for enabling the autopilot then user can enable it. We can not prevent user from setting that flag.

**Review by vinayakshnd — APPROVED** [review]
LGTM

**Pull request merged**